src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,106 +924,289 @@ namespace nbl::ext::imgui
-				auto memOffset = m_device->allocate(memReq, indexBuffer.get());
-				assert(memOffset.isValid());
-			}
+			struct MDI_PARAMS
+			{
+				std::array<MDI_SIZE_TYPE, MDI_COMPONENT_COUNT> bytesToFill = {}; //! used with MDI::E_BUFFER_CONTENT for elements
+				MDI_SIZE_TYPE totalByteSizeRequest = {}, //! sum of bytesToFill
+				drawCount = {};	//! amount of objects to draw for indirect call request
+			};
 
+			const MDI_PARAMS mdiParams = [&]()
 			{
-				auto vBinding = vertexBuffer->getBoundMemory();
-				auto iBinding = indexBuffer->getBoundMemory();
+				MDI_PARAMS params;
 
+				for (uint32_t i = 0; i < drawData->CmdListsCount; i++)
 				{
-					if (!vBinding.memory->map({ 0ull, vBinding.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
-						logger->log("Could not map device memory for vertex buffer data!", system::ILogger::ELL_WARNING);
-
-					assert(vBinding.memory->isCurrentlyMapped());
+					const ImDrawList* commandList = drawData->CmdLists[i];
+					params.bytesToFill[MDI::EBC_DRAW_INDIRECT_STRUCTURES] += commandList->CmdBuffer.Size * sizeof(VkDrawIndexedIndirectCommand);
+					params.bytesToFill[MDI::EBC_ELEMENT_STRUCTURES] += commandList->CmdBuffer.Size * sizeof(PerObjectData);
 				}
+				params.bytesToFill[MDI::EBC_VERTEX_BUFFERS] = drawData->TotalVtxCount * sizeof(ImDrawVert);
+				params.bytesToFill[MDI::EBC_INDEX_BUFFERS] = drawData->TotalIdxCount * sizeof(ImDrawIdx);
+
+				// calculate upper bound byte size limit for mdi buffer
+				params.totalByteSizeRequest = std::reduce(std::begin(params.bytesToFill), std::end(params.bytesToFill));
+				params.drawCount = params.bytesToFill[MDI::EBC_DRAW_INDIRECT_STRUCTURES] / sizeof(VkDrawIndexedIndirectCommand);
+
+				return std::move(params);
+			}();
 
+			std::array<bool, MDI_COMPONENT_COUNT> mdiBytesFilled;
+			std::fill(mdiBytesFilled.data(), mdiBytesFilled.data() + MDI_COMPONENT_COUNT, false);
+			std::array<MDI_SIZE_TYPE, MDI_COMPONENT_COUNT> mdiOffsets;
+			std::fill(mdiOffsets.data(), mdiOffsets.data() + MDI_COMPONENT_COUNT, INVALID_ADDRESS);
+
+			auto streamingBuffer = m_mdi.buffer;
+			{
+				auto binding = streamingBuffer->getBuffer()->getBoundMemory();
+				assert(binding.memory->isCurrentlyMapped());
+
+				auto* const mdiData = reinterpret_cast<uint8_t*>(streamingBuffer->getBufferPointer());
+
+				struct
 				{
-					if (!iBinding.memory->map({ 0ull, iBinding.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
-						logger->log("Could not map device memory for index buffer data!", system::ILogger::ELL_WARNING);
-				
-					assert(iBinding.memory->isCurrentlyMapped());
-				}
+					MDI_SIZE_TYPE offset = INVALID_ADDRESS,
+					multiAllocationSize = {};
+				} requestState;
 
-				auto* vertex_ptr = static_cast<ImDrawVert*>(vBinding.memory->getMappedPointer());
-				auto* index_ptr = static_cast<ImDrawIdx*>(iBinding.memory->getMappedPointer());
+				const auto start = std::chrono::steady_clock::now();
```

Comment:
take a timeout point (waitPt) from the outside, have it as the last argument and default it to `now()+milliseconds(1)`

Reply:
that way you can pass the same vairable to `multi_allocate`